components/butterfly/src/server/mod.rs

```diff
@@ -281,36 +317,220 @@ impl Server {
+        // of just how the logic goes. The value of the trace is really high, though, so we suck it
+        // for now.
+        let trace_member_id = String::from(member.get_id());
+        let trace_incarnation = member.get_incarnation();
+        let trace_health = health.clone();
+
+        if self.member_list.insert(member, health) || incremented_incarnation {
+            trace_it!(MEMBERSHIP: self,
+                      TraceKind::MemberUpdate,
+                      trace_member_id,
+                      trace_incarnation,
+                      trace_health);
+            self.rumor_list.insert(rk);
+        }
+    }
+
     /// Insert members from a list of received rumors.
-    pub fn insert_from_rumors(&self, members: Vec<(Member, Health)>) {
-        for (member, mut health) in members.into_iter() {
-            let mut incremented_incarnation = false;
-            let rk: RumorKey = RumorKey::from(&member);
-            if member.get_id() == self.member_id() {
-                if health != Health::Alive {
-                    let mut me = self.member.write().expect("Member lock is poisoned");
-                    let mut incarnation = me.get_incarnation();
-                    incarnation += 1;
-                    me.set_incarnation(incarnation);
-                    health = Health::Alive;
-                    incremented_incarnation = true;
+    pub fn insert_member_from_rumors(&self, members: Vec<(Member, Health)>) {
+        for (member, health) in members.into_iter() {
+            self.insert_member_from_rumor(member, health);
+        }
+    }
+
+    /// Insert a service rumor into the service store.
+    pub fn insert_service(&self, service: Service) {
+        let rk = RumorKey::from(&service);
+        if self.service_store.insert(service) {
+            self.rumor_list.insert(rk);
+        }
+    }
+
+    /// Get all the Member ID's who are present in a given service group.
+    pub fn get_electorate(&self, key: &str) -> Vec<String> {
+        let mut electorate = vec![];
+        self.service_store.with_rumors(key, |s| {
+            if self.member_list
+                .check_health_of_by_id(s.get_member_id(), Health::Alive) {
+                electorate.push(String::from(s.get_member_id()));
+            }
+        });
+        electorate
+    }
+
+    /// Check if a given service group has quorum to run an election.
+    ///
+    /// A given group has quorum if, from this servers perspective, it has an alive population that
+    /// is over 50%, and at least 3 members.
+    pub fn check_quorum(&self, key: &str) -> bool {
+        let electorate = self.get_electorate(key);
+
+        let total_population = self.service_store.len_for_key(key);
+        let alive_population = electorate.len();
+
+        if total_population < 3 {
+            info!("Quorum size: {}/3 - election cannot complete",
+                  total_population);
+            return false;
+        }
+
+        if total_population % 2 == 0 {
```

Comment:
If you don't want a warning, yes. If you have an even number of services in a group, and you want an election, you run the risk of both sides having quorum, which means you won't actually ever be able to complete an election.